icepack/models/shallow_ice.py

```diff
@@ -0,0 +1,187 @@
+# Copyright (C) 2020 by Jessica Badgeley <badgeley@uw.edu>
+#
+# This file is part of icepack.
+#
+# icepack is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# The full text of the license can be found in the file LICENSE in the
+# icepack source directory or at <http://www.gnu.org/licenses/>.
+
+import firedrake
+from firedrake import inner, grad, dx, ds
+from icepack.constants import (ice_density as ρ_I, water_density as ρ_W,
+                               gravity as g, glen_flow_law as n)
+from icepack.models.mass_transport import LaxWendroff
+from icepack.optimization import newton_search
+from icepack.utilities import add_kwarg_wrapper
+
+# A, the fluidity parameter, should be in kwargs even though it is required
+# because someone may want to use some other physics to describe A, and then
+# it would be easier to swap out.
+
+# Split viscosity_gravity into two parts (with the first plus sign). Use the
+# first part to compute the scale because it is always positive.
+
+def part_1(u):
+    r"""Return part 1 of the shallow ice action functional
+
+    Part 1 of the shallow ice action functional is
+
+    .. math::
+        E(u) = \frac{1}{2}\int_\Omega u\cdot u\; dx
+
+    Parameters
+    ----------
+    u : firedrake.Function
+        ice velocity
+
+    Returns
+    -------
+    firedrake.Form
+    """
+    return .5 * (inner(u,u))
+
+
+def part_2(u, h, s, A):
```

Comment:
I think a more descriptive name like `gravity` would be better here -- in principle someone might want to modify this part of the model and that'll help cue them as to what it is.